peps/pep-0782.rst

```diff
@@ -0,0 +1,348 @@
+
+Creating a Python :class:`bytes` object using
+``PyBytes_FromStringAndSize(NULL, size)`` and ``_PyBytes_Resize()``
+treats an immutable :class:`bytes` object as mutable. It goes against
+the principle that :class:`bytes` objects are immutable. It also creates
+an incomplete or "invalid" object since bytes are not initialized. In
+Python, a :class:`bytes` object should always have its bytes fully
+initialized.
+
+* `Avoid creating incomplete/invalid objects api-evolution#36
+  <https://github.com/capi-workgroup/api-evolution/issues/36>`_
+* `Disallow mutating immutable objects api-evolution#20
+  <https://github.com/capi-workgroup/api-evolution/issues/20>`_
+* `Disallow creation of incomplete/inconsistent objects problems#56
+  <https://github.com/capi-workgroup/problems/issues/56>`_
+
+Inefficient allocation strategy
+-------------------------------
+
+When creating a bytes string and the output size is unknown, one
+strategy is to allocate a short buffer and extend it (to the exact size)
+each time a larger write is needed.
+
+This strategy is inefficient because it requires to enlarge the buffer
+multiple timess. It's more efficient to overallocate the buffer the
+first time that a larger write is needed. It reduces the number of
+expensive ``realloc()`` operations which can imply a memory copy.
+
+
+Specification
+=============
+
+API
+---
+
+.. c:type:: PyBytesWriter
+
+   A Python :class:`bytes` writer instance created by
+   :c:func:`PyBytesWriter_Create`.
+
+   The instance must be destroyed by :c:func:`PyBytesWriter_Finish` or
+   :c:func:`PyBytesWriter_Discard`.
+
+Create, Finish, Discard
+^^^^^^^^^^^^^^^^^^^^^^^
+
+.. c:function:: PyBytesWriter* PyBytesWriter_Create(Py_ssize_t size)
+
+   Create a :c:type:`PyBytesWriter` to write *size* bytes.
+
+   If *size* is greater than zero, allocate *size* bytes for the
+   returned buffer.
+
+   On success, return ``0``.
+   On error, set an exception and return ``-1``.
+
+   *size* must be positive or zero.
+
+.. c:function:: PyObject* PyBytesWriter_Finish(PyBytesWriter *writer)
+
+   Finish a :c:type:`PyBytesWriter` created by
+   :c:func:`PyBytesWriter_Create`.
+
+   On success, return a Python :class:`bytes` object.
+   On error, set an exception and return ``NULL``.
+
+   The writer instance is invalid after the call in any case.
+
+.. c:function:: PyObject* PyBytesWriter_FinishWithSize(PyBytesWriter *writer, Py_ssize_t size)
+
+   Similar to :c:func:`PyBytesWriter_Finish`, but resize the writer
+   to *size* bytes before creating the :class:`bytes` object.
+
+.. c:function:: PyObject* PyBytesWriter_FinishWithPointer(PyBytesWriter *writer, void *buf)
+
+   Similar to :c:func:`PyBytesWriter_Finish`, but resize the writer
+   using *buf* pointer before creating the :class:`bytes` object.
+
+.. c:function:: void PyBytesWriter_Discard(PyBytesWriter *writer)
+
+   Discard a :c:type:`PyBytesWriter` created by :c:func:`PyBytesWriter_Create`.
+
+   The writer instance is invalid after the call.
```

Comment:
You should also mention that a NULL writer is allowed and that this function does nothing in this case (since your example is using this)